### DISCUSSION TITLE
Code generation - Solver interface

### DISCUSSION THREAD

**Discussion by jenkner** [casadi/casadi#3258] [Q&A]
I have written a Python 3.11 code to solve a Model Predictive Control (MPC) problem. The code works properly in Python, but when I try to generate the code using the following code:

```
MPC_func = opti.to_function('MPC', [x0, y0, psi0, delta0, v0], [x, y, psi, delta])
C = ca.CodeGenerator('MPC.c')
C.add(MPC_func)
C.generate()
```
CONSOLE OUTPUT:
`_CasADi - 2023-07-03 11:20:58 WARNING("The function "qpsol", which is of type "QpoasesInterface" cannot be code generated. The generation will proceed, but compilation of the code will not be possible.") [.../casadi/core/function_internal.cpp:2688]_`

I obtain the following line of code inside the generated code:

```
static int casadi_f6(const casadi_real** arg, casadi_real** res, casadi_int* iw, casadi_real* w, int mem) {
  #error Code generation not supported for QpoasesInterface
  return 0;
}
```

The error message clearly indicates that code generation is not supported for the QpoasesInterface. I have also tried using a different solver (ipopt), but I get a similar line of code.

According to the documentation, the interfaces to these solvers should be supported. Does anyone have an idea of what might be causing this issue?

**Comment by brunoml5** [comment]
Hi! Indeed, not all solvers are able to code generate.

For NLP, you need to use `sqpmethod` with a QP which supports code generation.  I tested many QP in Python 3.11, Casadi 3.6.3, Windows 11, VS compiler. The following QPs appears to work:

- `qrqp`: appears to work out of the box
- `osqp`: I believe code generation works, but the compilation failed for me. `osqp.h` not found

If you need code generation for MPC, take a look at [ACADOS](https://github.com/acados/acados). I believe their code generation interface is much more mature.

I hope it helps!

**Reply by jenkner** [reply]
Hi Bruno,

Thank you for your advice. I appreciate your suggestion. I will explore the alternative solver you mentioned, and I will also take a closer look at ACADOS.

**Reply by jgillis** [reply]
There is also 'highs' solver since 3.6 that is compatible with codegen..
